app/src/common/chromium/com/igalia/wolvic/browser/api/impl/DisplayImpl.java

```diff
@@ -9,42 +11,53 @@
 import com.igalia.wolvic.browser.api.WDisplay;
 import com.igalia.wolvic.browser.api.WResult;
 import com.igalia.wolvic.browser.api.WSession;
+import com.igalia.wolvic.utils.SystemUtils;
+
+import org.chromium.components.embedder_support.view.WolvicContentRenderView;
 
 public class DisplayImpl implements WDisplay {
-    @NonNull BrowserDisplay mDisplay;
+    static String LOGTAG = SystemUtils.createLogtag(DisplayImpl.class);
+
+    @NonNull private final BrowserDisplay mBrowserDisplay;
     @NonNull SessionImpl mSession;
     private int mWidth = 1;
     private Surface mSurface;
-    private boolean mIsFirstSurfaceChangedCall = true;
+    private WolvicContentRenderView mRenderView;
 
-    public DisplayImpl(@NonNull BrowserDisplay display, @NonNull SessionImpl session) {
-        mDisplay = display;
+    public DisplayImpl(
+            @NonNull BrowserDisplay browserDisplay,
+            @NonNull SessionImpl session,
+            WolvicContentRenderView renderView) {
+        mBrowserDisplay = browserDisplay;
+        mRenderView = renderView;
         mSession = session;
     }
 
-    @NonNull BrowserDisplay getBrowserDisplay() {
-        return mDisplay;
-    }
-
     @Override
     public void surfaceChanged(@NonNull Surface surface, int width, int height) {
         // Make sure we update the current active tab. We have an empty SessionImpl.setActive implementation
         // because in weblayer the browser window is selected from a pool. We can use the hint that
         // the session want's to be rendered into this surface to set it active too.
-        
-        if (mDisplay.getBrowser().getActiveTab() != mSession.mTab) {
-            mDisplay.getBrowser().setActiveTab(mSession.mTab);
-        }
+
         mWidth = width;
-        mDisplay.surfaceChanged(surface, width, height, mFirstCompositeRunnable);
-        if (!mIsFirstSurfaceChangedCall && mSurface != null && mSurface != surface) {
-            // Workaround for what it looks like a Chrome bug.
-            // When Surface is a different instance from previous one, chrome doesn't render
-            // correctly to it until navigation changes.
-            mSession.mTab.getNavigationController().reload();
-        }
-        mIsFirstSurfaceChangedCall = false;
         mSurface = surface;
+
+        try {
+            // Dispatch onSurfaceCreated
+            mRenderView.surfaceCreated(surface);
+
+            // Dispatch onSurfaceChanged
+            mRenderView.surfaceChanged(surface, width, height);
+
+            postDelayed((Runnable) () -> {
+                @Nullable WSession.ContentDelegate delegate = mSession.getContentDelegate();
+                if (delegate != null) {
+                    delegate.onFirstComposite(mSession);
+                }
+            }, 100);
```

Comment:
I believe we don't need this anymore. This has been addressed in the following PR: https://github.com/Igalia/wolvic/pull/514. I'm not sure if we want to address it here though, because we will be rebasing these in-progress CLs on top when this is merged.

Reply:
IMO, this statement should be cleaned here since it is for the weblayer implementation. Removed.